bee-trace/bee-trace/src/subscriber/layer/flamegraph.rs

```diff
@@ -0,0 +1,230 @@
+    where
+        S: Subscriber + for<'a> LookupSpan<'a>,
+    {
+        // Keep the current span, since we want to write the samples in the previous stack configuration,
+        // before this span was exited.
+        self.stack_string(id, ctx, false)
+    }
+
+    /// Writes a [`String`] that describes the stack of a given [`Span`](tracing::Span).
+    ///
+    /// This string is then directly used in the folded stack file to describe the number of samples in
+    /// the described stack.
+    fn stack_string<S>(&self, id: &span::Id, ctx: &Context<'_, S>, skip_current_span: bool) -> String
+    where
+        S: Subscriber + for<'a> LookupSpan<'a>,
+    {
+        let delta = self.time_since_last_event();
+        let mut stack_str = "all-spans".to_string();
+
+        // If the span is tracked but not written to the registry, something has gone very wrong.
+        let mut leaf_span = Some(ctx.span(id).expect("span is not in registry"));
+        if skip_current_span {
+            leaf_span = leaf_span.and_then(|span| span.parent());
+        }
+
+        if let Some(span) = leaf_span {
+            for span in span.scope().from_root() {
+                let location = match self
+                    .span_locations
+                    .read()
+                    .get(&span.id())
+                    .and_then(|location| location.as_ref())
+                {
+                    Some(location) => location.to_string(),
+                    None => "unknown".to_string(),
+                };
+
+                write!(stack_str, "; {}", location).expect("writing to String should never fail");
+            }
+        }
+
+        let _ = write!(stack_str, " {}", delta.as_micros());
+        stack_str
+    }
+
+    /// Returns the time since the last event within this [`Span`](tracing::Span) stack, and updates the
+    /// last event time to the current time.
+    fn time_since_last_event(&self) -> Duration {
+        let now = Instant::now();
+
+        let last_event = LAST_EVENT.with(|time| {
+            let last_event = time.get();
+            time.set(now);
+            last_event
+        });
+
+        now - last_event
+    }
+}
+
+impl<S> Layer<S> for FlamegraphLayer
+where
+    S: Subscriber + for<'a> LookupSpan<'a>,
+{
+    fn register_callsite(&self, metadata: &'static Metadata<'static>) -> subscriber::Interest {
+        match (metadata.name(), metadata.target()) {
+            ("runtime.spawn", _) | ("task", "tokio::task") | (_, "bee::observe") => {
+                self.callsites.write().insert(metadata.callsite());
+            }
+            (_, _) => {}
+        }
+
+        // Ensure any downstream layers have access to all spans and events.
+        subscriber::Interest::always()
+    }
+
+    fn new_span(&self, attrs: &span::Attributes<'_>, id: &span::Id, _ctx: Context<'_, S>) {
+        if !self.is_tracked_callsite(&attrs.metadata().callsite()) {
```

Comment:
can't this be:
```rust
       if self.is_tracked_callsite(&attrs.metadata().callsite()) {
           let location = SpanLocation::from_attributes(attrs);
           self.span_locations.write().insert(id.clone(), location);
       }
```

Reply:
It can, but I prefer the early return way :trollface: Not that bothered about it though.

Reply:
@thibault-martinez we have an instance of the CAT theorem but with `A = Adam` fix plz :trollface:

Reply:
I have a tendency of going for the least verbose option if it doesn't degrade readability, and I think here it wouldn't so I would agree with Christian

Reply:
Sure, I'll change them 👍